// @likec4/core/dist/styles/index.mjs
import { t as t$5 } from '../shared/core.q10OP9oJ.mjs';
import { t as t$3, c as compareNatural } from '../shared/core.3ZhnKAlZ.mjs';
import { t as t$2, e as e$1 } from '../shared/core.Coi_LOUZ.mjs';
import { e } from '../shared/core.BKlholso.mjs';
import { t as t$6 } from '../shared/core.CbT8JVYL.mjs';
import { t as t$1 } from '../shared/core.BsSXS-_R.mjs';
import { t as t$4 } from '../shared/core.BkVA6gjl.mjs';
import { a as nonNullable } from '../shared/core.D4npX2q8.mjs';
export { B as BorderStyles, E as ElementShapes, L as LikeC4Styles, R as RelationshipArrowTypes, S as Sizes, T as ThemeColors, c as computeColorValues, d as defaultTheme, b as isCustomColor, i as isThemeColor, s as styleDefaults } from '../shared/core.DGcIbsj4.mjs';

function t(...t){return t$1(n,t)}function n(e,t,n){let r={};for(let[i,a]of e.entries()){let o=t(a,i,e);r[o]=n(a,i,e);}return r}

const DefaultTagColors = [
  "tomato",
  "grass",
  "blue",
  "ruby",
  "orange",
  "indigo",
  "pink",
  "teal",
  "purple",
  "amber",
  "crimson",
  "red",
  "lime",
  "yellow",
  "violet"
];
function assignTagColors(tags) {
  const tagsWithColors = [];
  const tagsWithoutColors = [];
  for (const [tag, spec] of t$2(tags)) {
    if (e(spec.color)) {
      tagsWithColors.push({
        tag,
        spec: {
          color: spec.color
        }
      });
    } else {
      tagsWithoutColors.push(tag);
    }
  }
  return t$3(
    tagsWithoutColors,
    t$4(compareNatural),
    t$6((tag, idx) => {
      const color = nonNullable(DefaultTagColors[idx % DefaultTagColors.length]);
      return {
        tag,
        spec: {
          color
        }
      };
    }),
    t$5(tagsWithColors),
    t$4((a, b) => compareNatural(a.tag, b.tag)),
    t(e$1("tag"), e$1("spec"))
  );
}

export { DefaultTagColors, assignTagColors };
