// @likec4/core/dist/types/expression-model.mjs
import { FqnRef } from './fqnRef.mjs';

var ModelFqnExpr;
((ModelFqnExpr2) => {
  function isWildcard(expr) {
    return "wildcard" in expr && expr.wildcard === true;
  }
  ModelFqnExpr2.isWildcard = isWildcard;
  function isModelRef(ref) {
    return "ref" in ref && FqnRef.isModelRef(ref.ref);
  }
  ModelFqnExpr2.isModelRef = isModelRef;
  function isElementKindExpr(expr) {
    return "elementKind" in expr && "isEqual" in expr;
  }
  ModelFqnExpr2.isElementKindExpr = isElementKindExpr;
  function isElementTagExpr(expr) {
    return "elementTag" in expr && "isEqual" in expr;
  }
  ModelFqnExpr2.isElementTagExpr = isElementTagExpr;
  function isWhere(expr) {
    return "where" in expr && is(expr.where.expr);
  }
  ModelFqnExpr2.isWhere = isWhere;
  function isCustom(expr) {
    return "custom" in expr && (is(expr.custom.expr) || isWhere(expr.custom.expr));
  }
  ModelFqnExpr2.isCustom = isCustom;
  function is(expr) {
    return isWildcard(expr) || isModelRef(expr) || isElementKindExpr(expr) || isElementTagExpr(expr);
  }
  ModelFqnExpr2.is = is;
  function unwrap(expr) {
    if (isCustom(expr)) {
      expr = expr.custom.expr;
    }
    if (isWhere(expr)) {
      expr = expr.where.expr;
    }
    return expr;
  }
  ModelFqnExpr2.unwrap = unwrap;
})(ModelFqnExpr || (ModelFqnExpr = {}));
var ModelRelationExpr;
((ModelRelationExpr2) => {
  function isDirect(expr) {
    return "source" in expr && "target" in expr;
  }
  ModelRelationExpr2.isDirect = isDirect;
  function isIncoming(expr) {
    return "incoming" in expr;
  }
  ModelRelationExpr2.isIncoming = isIncoming;
  function isOutgoing(expr) {
    return "outgoing" in expr;
  }
  ModelRelationExpr2.isOutgoing = isOutgoing;
  function isInOut(expr) {
    return "inout" in expr;
  }
  ModelRelationExpr2.isInOut = isInOut;
  function isWhere(expr) {
    return "where" in expr && (isDirect(expr.where.expr) || isIncoming(expr.where.expr) || isOutgoing(expr.where.expr) || isInOut(expr.where.expr));
  }
  ModelRelationExpr2.isWhere = isWhere;
  function isCustom(expr) {
    return "customRelation" in expr;
  }
  ModelRelationExpr2.isCustom = isCustom;
  function is(expr) {
    return isDirect(expr) || isIncoming(expr) || isOutgoing(expr) || isInOut(expr);
  }
  ModelRelationExpr2.is = is;
  function unwrap(expr) {
    if (isCustom(expr)) {
      expr = expr.customRelation.expr;
    }
    if (isWhere(expr)) {
      expr = expr.where.expr;
    }
    return expr;
  }
  ModelRelationExpr2.unwrap = unwrap;
})(ModelRelationExpr || (ModelRelationExpr = {}));
var ModelExpression;
((ModelExpression2) => {
  function isWhere(expr) {
    return "where" in expr;
  }
  ModelExpression2.isWhere = isWhere;
  function isRelationWhere(expr) {
    return ModelRelationExpr.isWhere(expr);
  }
  ModelExpression2.isRelationWhere = isRelationWhere;
  function isFqnExprWhere(expr) {
    return ModelFqnExpr.isWhere(expr);
  }
  ModelExpression2.isFqnExprWhere = isFqnExprWhere;
  function isFqnExpr(expr) {
    return ModelFqnExpr.is(expr) || ModelFqnExpr.isWhere(expr) || ModelFqnExpr.isCustom(expr);
  }
  ModelExpression2.isFqnExpr = isFqnExpr;
  function isRelationExpr(expr) {
    return ModelRelationExpr.is(expr) || ModelRelationExpr.isWhere(expr) || ModelRelationExpr.isCustom(expr);
  }
  ModelExpression2.isRelationExpr = isRelationExpr;
})(ModelExpression || (ModelExpression = {}));

export { ModelExpression, ModelFqnExpr, ModelRelationExpr };
