// @likec4/core/dist/types/expression.mjs
import { FqnRef } from './fqnRef.mjs';

var FqnExpr;
((FqnExpr2) => {
  function isWildcard(expr) {
    return "wildcard" in expr && expr.wildcard === true;
  }
  FqnExpr2.isWildcard = isWildcard;
  function isModelRef(ref) {
    return "ref" in ref && FqnRef.isModelRef(ref.ref);
  }
  FqnExpr2.isModelRef = isModelRef;
  function isDeploymentRef(expr) {
    return "ref" in expr && FqnRef.isDeploymentRef(expr.ref);
  }
  FqnExpr2.isDeploymentRef = isDeploymentRef;
  function isElementKindExpr(expr) {
    return "elementKind" in expr && "isEqual" in expr;
  }
  FqnExpr2.isElementKindExpr = isElementKindExpr;
  function isElementTagExpr(expr) {
    return "elementTag" in expr && "isEqual" in expr;
  }
  FqnExpr2.isElementTagExpr = isElementTagExpr;
  function isWhere(expr) {
    return "where" in expr && is(expr.where.expr);
  }
  FqnExpr2.isWhere = isWhere;
  function isCustom(expr) {
    return "custom" in expr && (is(expr.custom.expr) || isWhere(expr.custom.expr));
  }
  FqnExpr2.isCustom = isCustom;
  function is(expr) {
    return isWildcard(expr) || isModelRef(expr) || isDeploymentRef(expr) || isElementKindExpr(expr) || isElementTagExpr(expr);
  }
  FqnExpr2.is = is;
  function unwrap(expr) {
    if (isCustom(expr)) {
      expr = expr.custom.expr;
    }
    if (isWhere(expr)) {
      expr = expr.where.expr;
    }
    return expr;
  }
  FqnExpr2.unwrap = unwrap;
})(FqnExpr || (FqnExpr = {}));
var RelationExpr;
((RelationExpr2) => {
  function isDirect(expr) {
    return "source" in expr && "target" in expr;
  }
  RelationExpr2.isDirect = isDirect;
  function isIncoming(expr) {
    return "incoming" in expr;
  }
  RelationExpr2.isIncoming = isIncoming;
  function isOutgoing(expr) {
    return "outgoing" in expr;
  }
  RelationExpr2.isOutgoing = isOutgoing;
  function isInOut(expr) {
    return "inout" in expr;
  }
  RelationExpr2.isInOut = isInOut;
  function isWhere(expr) {
    return "where" in expr && (isDirect(expr.where.expr) || isIncoming(expr.where.expr) || isOutgoing(expr.where.expr) || isInOut(expr.where.expr));
  }
  RelationExpr2.isWhere = isWhere;
  function isCustom(expr) {
    return "customRelation" in expr;
  }
  RelationExpr2.isCustom = isCustom;
  function is(expr) {
    return isDirect(expr) || isIncoming(expr) || isOutgoing(expr) || isInOut(expr);
  }
  RelationExpr2.is = is;
  function unwrap(expr) {
    if (isCustom(expr)) {
      expr = expr.customRelation.expr;
    }
    if (isWhere(expr)) {
      expr = expr.where.expr;
    }
    return expr;
  }
  RelationExpr2.unwrap = unwrap;
})(RelationExpr || (RelationExpr = {}));
var Expression;
((Expression2) => {
  function isWhere(expr) {
    return "where" in expr;
  }
  Expression2.isWhere = isWhere;
  function isRelationWhere(expr) {
    return RelationExpr.isWhere(expr);
  }
  Expression2.isRelationWhere = isRelationWhere;
  function isFqnExprWhere(expr) {
    return FqnExpr.isWhere(expr);
  }
  Expression2.isFqnExprWhere = isFqnExprWhere;
  function isFqnExpr(expr) {
    return FqnExpr.is(expr) || FqnExpr.isWhere(expr) || FqnExpr.isCustom(expr);
  }
  Expression2.isFqnExpr = isFqnExpr;
  function isRelation(expr) {
    return RelationExpr.is(expr) || RelationExpr.isWhere(expr) || RelationExpr.isCustom(expr);
  }
  Expression2.isRelation = isRelation;
})(Expression || (Expression = {}));

export { Expression, FqnExpr, RelationExpr };
