// @likec4/core/dist/types/fqnRef.mjs
import { e } from '../shared/core.CqAJJpK_.mjs';
import { a as GlobalFqn } from '../shared/core.D0WZutNd.mjs';

var FqnRef;
((FqnRef2) => {
  function isElementRef(ref) {
    return "model" in ref && !("project" in ref);
  }
  FqnRef2.isElementRef = isElementRef;
  function isImportRef(ref) {
    return "project" in ref && "model" in ref;
  }
  FqnRef2.isImportRef = isImportRef;
  function flatten(ref) {
    if (e(ref)) {
      throw new Error(`Expected FqnRef, got: "${ref}"`);
    }
    if (isImportRef(ref)) {
      return GlobalFqn(ref.project, ref.model);
    }
    if (isElementRef(ref)) {
      return ref.model;
    }
    throw new Error("Expected FqnRef.ModelRef or FqnRef.ImportRef");
  }
  FqnRef2.flatten = flatten;
  function isModelRef(ref) {
    return isElementRef(ref) || isImportRef(ref);
  }
  FqnRef2.isModelRef = isModelRef;
  function isInsideInstanceRef(ref) {
    return "deployment" in ref && "element" in ref;
  }
  FqnRef2.isInsideInstanceRef = isInsideInstanceRef;
  function isDeploymentElementRef(ref) {
    return "deployment" in ref && !("element" in ref);
  }
  FqnRef2.isDeploymentElementRef = isDeploymentElementRef;
  function isDeploymentRef(ref) {
    return isDeploymentElementRef(ref) || isInsideInstanceRef(ref);
  }
  FqnRef2.isDeploymentRef = isDeploymentRef;
})(FqnRef || (FqnRef = {}));

export { FqnRef };
