// @likec4/core/dist/utils/index.mjs
export { i as invariant, a as nonNullable, n as nonexhaustive } from '../shared/core.D4npX2q8.mjs';
export { S as Stack, c as commonHead, a as compareRelations, o as objectHash } from '../shared/core.btEq3i3f.mjs';
export { c as compareNatural, b as compareNaturalHierarchically } from '../shared/core.3ZhnKAlZ.mjs';
export { a as ancestorsFqn, g as commonAncestor, j as compareByFqnHierarchically, k as compareFqnHierarchically, h as hasProp, l as hierarchyDistance, m as hierarchyLevel, i as isAncestor, q as isAnyOf, b as isDescendantOf, c as isNonEmptyArray, d as isSameHierarchy, e as isString, n as nameFromFqn, p as parentFqn, o as sortByFqnHierarchically, s as sortNaturalByFqn, f as sortParentsFirst } from '../shared/core.CdNCAwq7.mjs';
export { g as getOrCreate } from '../shared/core.DcS-0zys.mjs';
export { m as memoizeProp } from '../shared/core.CLSZtnFh.mjs';
export { d as difference, e as equalsSet, i as intersection, s as symmetricDifference, u as union } from '../shared/core.DbfO90vF.mjs';
export { s as stringHash } from '../shared/core.oSV6SRfk.mjs';
export { b as BiMap, d as DefaultWeakMap, l as LinkedList, m as MultiMap, q as Queue } from '../shared/core.BNN-3vPH.mjs';
export { D as DefaultMap } from '../shared/core.BbE4y-yl.mjs';
export { i as ifilter } from '../shared/core.DRJma0Ol.mjs';
export { i as ifind } from '../shared/core.CmPXpJ8l.mjs';
export { ifirst, iflatMap, ireduce } from './iterable/index.mjs';
export { i as iflat, a as iunique } from '../shared/core.C4x8aaWa.mjs';
export { i as ihead } from '../shared/core.uFk_o9X6.mjs';
export { i as imap, a as toArray, t as toSet } from '../shared/core.C3JowhaZ.mjs';
export { i as isIterable } from '../shared/core.RfnaEmgA.mjs';
export { i as isome } from '../shared/core.FZ305VYd.mjs';
export { m as markdownToHtml, a as markdownToText } from '../shared/core.Bdy3hhTc.mjs';

function e(e,t){let n=Math.ceil(e),r=Math.floor(t);if(r<n)throw RangeError(`randomInteger: The range [${e.toString()},${t.toString()}] contains no integer`);return Math.floor(Math.random()*(r-n+1)+n)}

const DELAY = "LIKEC4_DELAY";
function delay(...args) {
  let ms = 100;
  if (args.length === 2) {
    ms = e(args[0], args[1]);
  } else if (args.length === 1) {
    ms = args[0];
  }
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve(DELAY);
    }, ms ?? 100);
  });
}
function promiseNextTick() {
  return Promise.resolve().then(() => void 0);
}
function onNextTick(fn) {
  void Promise.resolve().then(fn);
}

export { delay, onNextTick, promiseNextTick };
