// @likec4/core/dist/shared/core.DH7TXqXf.mjs
import { e as e$1 } from './core.BKlholso.mjs';
import { t as t$2 } from './core.BsSXS-_R.mjs';
import { e as e$2 } from './core.BoYfZOBm.mjs';
import { e as e$3 } from './core.CqAJJpK_.mjs';
import { n as nonexhaustive } from './core.D4npX2q8.mjs';

function t$1(...t){return t$2(n$1,t)}const n$1=(e,t)=>t.every(t=>t(e));

function t(...t){return t$2(n,t)}const n=(e,t)=>t.some(t=>t(e));

function e(e){return e==null}

function isOnStage(value, stage) {
  return value[_stage] === stage;
}
const _stage = "_stage";
const _type = "_type";
const _layout = "_layout";

function isDeploymentNode(el) {
  return e$1(el.kind) && !e$1(el.element);
}
function isDeployedInstance(el) {
  return e$1(el.element) && !e$1(el.kind);
}

function isTagEqual(operator) {
  return "tag" in operator;
}
function isKindEqual(operator) {
  return "kind" in operator;
}
function isParticipantOperator(operator) {
  return "participant" in operator;
}
function isNotOperator(operator) {
  return "not" in operator;
}
function isAndOperator(operator) {
  return "and" in operator;
}
function isOrOperator(operator) {
  return "or" in operator;
}
function whereOperatorAsPredicate(operator) {
  switch (true) {
    case isParticipantOperator(operator): {
      const participant = operator.participant;
      const participantPredicate = whereOperatorAsPredicate(operator.operator);
      return participantIs(participant, participantPredicate);
    }
    case isTagEqual(operator): {
      if (e$3(operator.tag) || "eq" in operator.tag) {
        const tag2 = e$3(operator.tag) ? operator.tag : operator.tag.eq;
        return (value) => {
          return Array.isArray(value.tags) && value.tags.includes(tag2);
        };
      }
      const tag = operator.tag.neq;
      return (value) => {
        return !Array.isArray(value.tags) || !value.tags.includes(tag);
      };
    }
    case isKindEqual(operator): {
      if (e$3(operator.kind) || "eq" in operator.kind) {
        const kind2 = e$3(operator.kind) ? operator.kind : operator.kind.eq;
        return (value) => {
          return value.kind === kind2;
        };
      }
      const kind = operator.kind.neq;
      return (value) => {
        return e(value.kind) || value.kind !== kind;
      };
    }
    case isNotOperator(operator): {
      const predicate = whereOperatorAsPredicate(operator.not);
      return e$2(predicate);
    }
    case isAndOperator(operator): {
      const predicates = operator.and.map(whereOperatorAsPredicate);
      return t$1(predicates);
    }
    case isOrOperator(operator): {
      const predicates = operator.or.map(whereOperatorAsPredicate);
      return t(predicates);
    }
    default:
      nonexhaustive(operator);
  }
}
function participantIs(participant, predicate) {
  return (value) => {
    if (!value.source || !value.target) {
      return false;
    }
    switch (participant) {
      case "source": {
        return predicate(value.source);
      }
      case "target": {
        return predicate(value.target);
      }
    }
  };
}

function isViewRulePredicate(rule) {
  return "include" in rule || "exclude" in rule;
}
function isViewRuleStyle(rule) {
  return "targets" in rule && "style" in rule;
}
function isComputedView(view) {
  return view._stage === "computed";
}
function isDiagramView(view) {
  return view._stage === "layouted";
}
function isElementView(view) {
  return view._type === "element";
}
function isScopedElementView(view) {
  return isElementView(view) && e$1(view.viewOf);
}
function isExtendsElementView(view) {
  return isElementView(view) && e$1(view.extends);
}
function isDeploymentView(view) {
  return view._type === "deployment";
}
function isDynamicView(view) {
  return view._type === "dynamic";
}

export { _stage as _, _type as a, _layout as b, isDeploymentNode as c, isDeployedInstance as d, isTagEqual as e, isKindEqual as f, isParticipantOperator as g, isNotOperator as h, isOnStage as i, isAndOperator as j, isOrOperator as k, isViewRulePredicate as l, isViewRuleStyle as m, isComputedView as n, isDiagramView as o, isElementView as p, isScopedElementView as q, isExtendsElementView as r, isDeploymentView as s, isDynamicView as t, e as u, t as v, whereOperatorAsPredicate as w };
