// @likec4/core/dist/shared/core.D0WZutNd.mjs
import { e } from './core.CqAJJpK_.mjs';
import { e as e$1 } from './core.BKlholso.mjs';
import { i as invariant } from './core.D4npX2q8.mjs';

function ProjectId(name) {
  return name;
}
function flattenMarkdownOrString(value) {
  if (value === null || value === void 0) {
    return null;
  }
  const content = e(value) ? value : value.txt ?? value.md;
  return e$1(content?.trim()) ? content : null;
}
const NoneIcon = "none";
function Fqn(name, parent) {
  return parent ? parent + "." + name : name;
}
const GroupElementKind = "@group";
function isGroupElementKind(v) {
  return v.kind === GroupElementKind;
}
function DeploymentFqn(name, parent) {
  return parent ? parent + "." + name : name;
}
function ViewId(id) {
  return id;
}
function RelationId(id) {
  return id;
}
function GlobalFqn(projectId, name) {
  invariant(typeof projectId === "string" && projectId != "");
  return "@" + projectId + "." + name;
}
function isGlobalFqn(fqn) {
  return fqn.startsWith("@");
}
function splitGlobalFqn(fqn) {
  if (!fqn.startsWith("@")) {
    return [null, fqn];
  }
  const firstDot = fqn.indexOf(".");
  if (firstDot < 2) {
    throw new Error("Invalid global FQN");
  }
  const projectId = fqn.slice(1, firstDot);
  const name = fqn.slice(firstDot + 1);
  return [projectId, name];
}
function NodeId(id) {
  return id;
}
function EdgeId(id) {
  return id;
}
function stepEdgeId(step, parallelStep) {
  const id = `step-${String(step).padStart(2, "0")}`;
  return parallelStep ? `${id}.${parallelStep}` : id;
}
const StepEdgeKind = "@step";
function isStepEdgeId(id) {
  return id.startsWith("step-");
}
function extractStep(id) {
  if (!isStepEdgeId(id)) {
    throw new Error(`Invalid step edge id: ${id}`);
  }
  return parseFloat(id.slice("step-".length));
}

const scalar = {
  __proto__: null,
  DeploymentFqn: DeploymentFqn,
  EdgeId: EdgeId,
  Fqn: Fqn,
  GlobalFqn: GlobalFqn,
  GroupElementKind: GroupElementKind,
  NodeId: NodeId,
  NoneIcon: NoneIcon,
  ProjectId: ProjectId,
  RelationId: RelationId,
  StepEdgeKind: StepEdgeKind,
  ViewId: ViewId,
  extractStep: extractStep,
  flattenMarkdownOrString: flattenMarkdownOrString,
  isGlobalFqn: isGlobalFqn,
  isGroupElementKind: isGroupElementKind,
  isStepEdgeId: isStepEdgeId,
  splitGlobalFqn: splitGlobalFqn,
  stepEdgeId: stepEdgeId
};

export { DeploymentFqn as D, EdgeId as E, Fqn as F, GroupElementKind as G, NoneIcon as N, ProjectId as P, RelationId as R, StepEdgeKind as S, ViewId as V, GlobalFqn as a, isGlobalFqn as b, splitGlobalFqn as c, NodeId as d, stepEdgeId as e, flattenMarkdownOrString as f, isStepEdgeId as g, extractStep as h, isGroupElementKind as i, scalar as s };
