// @likec4/core/dist/shared/core.om36dXXW.mjs
import { g as isStepEdgeId } from './core.D0WZutNd.mjs';

function e(e){return Array.isArray(e)}

function isViewRuleGlobalStyle(rule) {
  return "styleId" in rule;
}
function isViewRuleGlobalPredicateRef(rule) {
  return "predicateId" in rule;
}
function isAutoLayoutDirection(autoLayout) {
  return autoLayout === "TB" || autoLayout === "BT" || autoLayout === "LR" || autoLayout === "RL";
}
function isViewRuleAutoLayout(rule) {
  return "direction" in rule;
}

function getParallelStepsPrefix(id) {
  if (isStepEdgeId(id) && id.includes(".")) {
    return id.slice(0, id.indexOf(".") + 1);
  }
  return null;
}
function isDynamicStep(step) {
  return !!step && !("__series" in step || "__parallel" in step);
}
function isDynamicStepsParallel(step) {
  return !!step && "__parallel" in step && e(step.__parallel);
}
function isDynamicStepsSeries(step) {
  return !!step && "__series" in step && e(step.__series);
}

function isViewRuleGroup(rule) {
  return "title" in rule && "groupRules" in rule && Array.isArray(rule.groupRules);
}
function isViewRuleRank(rule) {
  return "rank" in rule && Array.isArray(rule.targets);
}

export { isViewRuleGlobalPredicateRef as a, isAutoLayoutDirection as b, isViewRuleAutoLayout as c, isDynamicStep as d, isDynamicStepsParallel as e, isDynamicStepsSeries as f, getParallelStepsPrefix as g, isViewRuleGroup as h, isViewRuleGlobalStyle as i, isViewRuleRank as j, e as k };
